### ISSUE TITLE
how to prevent "dhcpcd %iface%" from exiting if no IP received?

### ISSUE THREAD

**Issue by perkelix** [NetworkConfiguration/dhcpcd#423]
Not a bug, as much as a question.

When launched via `dhcpcd %iface%`, the binary seemingly exits if no answer was received. Is there any command option that can be passed to make the binary fork to the background if the timeout has been reached and keep on waiting for a reply?

The` --timeout` option seems like it might do what we need, but the manual page's phrasing is ambiguous.

I'm asking because, if the binary exits, it makes `ifup` fail irreversibly. At that point, only a host to which the administrator has physical access can be logged onto to have a manual attempt at acquiring an IP, while remote hosts are essentially lost due to the binary exiting.

**Comment by rsmarples** [comment]
Is this a dupe of #413?

**Comment by perkelix** [comment]
No. The other one wants to wait for an error before exiting. This one wants to not exit and instead fork to the background to wait for an IP.

**Comment by ColinMcInnes** [comment]
Could you use managed mode? Then everything happens in the background, and it keeps trying to infinity.

The only downside is that timeout doesn't do anything at that point, if you're looking for feedback that the timeout was observed.

**Comment by perkelix** [comment]
I'm returning to this because it's a key difference with dhclient that's gonna bite people pretty hard. On dhclient, in addition to a timeout, there's a retry interval. Basically, dhclient never exits; it merely goes to sleep and tries again later. By contrast, dhcpcd becomes non-recoverable without physical access to the host, if it fails to acquire an IP before timeout.

**Comment by ColinMcInnes** [comment]
If you want it to keep retrying, why not use managed mode? I wouldn't recommend trying to shoehorn interactive mode on a headless host.

**Comment by perkelix** [comment]
Because it's incompatible with the interface-oriented approach of ifupdown.

**Comment by ColinMcInnes** [comment]
If it forks to the background after the timeout, how would it communicate the eventual success if there is one?

In my custom build for my project I added more conditions that trigger exit hooks so I can react to things like how many discovers were sent, or whether the backoff has reached max, etc. Could that be something that would be helpful? I wouldn't want to spam the base system so I'd have to gate it behind some kind of config. But it has definitely proven useful to communicate what state dhcpcd is in to the rest of the system.

**Comment by rsmarples** [comment]
> Because it's incompatible with the interface-oriented approach of ifupdown.

It's not compatible with the reliance on the exit code approach of ifupdown.

Consider this scenario:
1 dhclient eth0
2 dhclient exits code 0 to show success! Yeah, we got an ip address
3 ifup carries on it's merry way on the path to great success!
4 dhclient renews after 20 seconds, fails, rebinds after 10 more seconds, fails and rips the IP address out from under you
5 how is the host now recoverable without physical access
6 ???
7 profit
8 realise there is no profit in exit codes when state changes badly

dhcpcd has been event driven since dhcpcd-4 at least and forces this model via dhcpcd-run-hooks(8) which is a similar approach to  dhclient-script. It's just that dhcpcd is more in your face with it because the hooks model is built around it whereas dhclient is more monolithic and per OS.

Now it's not all doom and gloom! Remember that IPv4LL is enabled by default so if dhcpcd does fail to get an IPv4 address from DHCP, it's still reachable on the local LAN via IPv4LL *and* by IPv6 LL address. You configure mDNS correctly and you don't even need a DHCP client on a well formed IPv6 network to access on the LAN.

**Comment by perkelix** [comment]
It's not relying on the exit code. It's relying on dhclient retrying at intervals if it's lost its IP or failed to acquire one. dhcpcd's IPv4LL fallback really is no substitute for that, especially not in a corporate network where IPs are typically assigned per MAC address.

**Comment by perkelix** [comment]
@rsmarples Unless I'm mistaken, doing what Debian expects for a drop-in replacement to `dhclient` would require the following options (or their equivalent command line flags)? i.e. without timeout=0 we won't keep on retrying, without noipv4ll we'll fallback onto Bonjour instead of retrying.

```
background
timeout=0
noipv4ll
```

Is this what @DanielG was suggesting as the default?

**Comment by ColinMcInnes** [comment]
@perkelix 
I think the issue of step 4 in Roy's example is still a concern. Relying on exit codes and then backgrounding means if a future renew fails, there's no feedback to the system that the renew failed.

Does ifupdown react to failed renews via dhclient? How does it know if dhclient was moved to the background by that point?

**Comment by perkelix** [comment]
@ColinMcInnes dhclient simply never gives up. It has both a timeout and a retry interval. If a timeout occurs, it sleeps for a while and tries again. This provides a good foolproof against loss of network or other kinds of failures to acquire an IP. One can rest assured that a host running dhclient will reconnect as soon as feasible and become reachable again.

With dhcpcd, @rsmarples bets on IPV4ll providing a sufficient backup strategy. In practice, it doesn't quite work as well, since the host no longer is reachable on the usual IP space it's expected to be found.

**Comment by ColinMcInnes** [comment]
dhcpcd should never give up either, and keep trying DISCOVER even after it falls back to LL, so if it eventually gets through, the device would get an IP address from the server.

Have you found this is not the case?

**Comment by perkelix** [comment]
> dhcpcd should never give up either, and keep trying DISCOVER even after it falls back to LL, so if it eventually gets through, the device would get an IP address from the server.
> 
> Have you found this is not the case?

@ColinMcInnes It has happened to me once. My DHCP server was rebooting after a kernel upgrade and my testing host running dhcpcd via ifupdown failed to acquire an IP before timeout, so it exited. Meanwhile, my other host running dhcpcd as a daemon patiently waited for the DHCP server to reboot and eventually got an IP.

**Comment by rsmarples** [comment]
@perkelix those commits should solve your problem and should also apply cleanly to dhcpcd-10.1.x. Let me know!
NOTE: This is a change to the historical behaviour of dhcpcd since v1 - is that a concern for Debian? That's why the dhcpcd5 package was initially created because I retired an option that didn't make sense anymore.

I think I just misunderstood what needed doing in this whole ticket. Sorry!

**Comment by perkelix** [comment]
> [@perkelix](https://github.com/perkelix) those commits should solve your problem and should also apply cleanly to dhcpcd-10.1.x. Let me know! NOTE: This is a change to the historical behaviour of dhcpcd since v1 - is that a concern for Debian? That's why the dhcpcd5 package was initially created because I retired an option that didn't make sense anymore.
> 
> I think I just misunderstood what needed doing in this whole ticket. Sorry!

Thanks for the followup. These don't apply cleanly to what we have in Trixie:

```
Applying patch 6acf895a66b143c42308777e085c1511cdebe5e0.patch
patching file src/dhcpcd.c
patching file src/if-options.h
Hunk #1 FAILED at 63.
Hunk #2 succeeded at 85 (offset 4 lines).
1 out of 2 hunks FAILED -- rejects in file src/if-options.h
Patch 6acf895a66b143c42308777e085c1511cdebe5e0.patch does not apply (enforce with -f)

```

```
Applying patch 5db90127b1c3128480a52559a9cdbd4949a0fed4.patch
patching file src/dhcpcd.8.in
Hunk #1 succeeded at 523 (offset 1 line).
patching file src/dhcpcd.c

Now at patch 5db90127b1c3128480a52559a9cdbd4949a0fed4.patch
```

**Comment by rsmarples** [comment]
Excellent! So just let me know if it works for you now.

**Comment by perkelix** [comment]
Actually, I had to remove the fragment that changes DHCPCD_DEBUG to a comment, otherwise it won't build.

**Comment by perkelix** [comment]
PS: this is the modified commit I used: https://salsa.debian.org/debian/dhcpcd/-/commit/11725d7d995fb731e80b97b045157e47dfa49a9c

**Comment by perkelix** [comment]
@rsmarples Could we have this into master as well?
